Add a deployment guide for astro (#1315)
clerk/clerk-docs: docs/deployments/deploy-astro.mdx

---
description: Learn how to deploy an Astro app to production with Clerk.
---

# Deploy an Astro app to production

> [!WARNING]
> This guide assumes that you have already installed Clerk in your application locally and are ready to deploy. If you haven't installed Clerk yet, please see our [guide to installing Clerk in your Astro application](/docs/quickstarts/astro) first.

This guide will walk you through the steps to deploy your Astro app to production. 

## Deploy your Clerk app to production

Follow the [guide on deploying your Clerk app to production](/docs/deployments/overview).

> ## Choose an adapter

Author: ```suggestion
## Deploy to a hosting platform
```

> ## Choose an adapter

Reviewer: ```suggestion
## Deploy to a hosting platform
```

> We recommend you choose one of the [official adapters](https://docs.astro.build/en/guides/server-side-rendering/#official-adapters) Astro offers. Below we are highlighting useful information for each one of them.
> 
> - [@astro/cloudflare](https://docs.astro.build/en/guides/integrations-guide/cloudflare/): ✅ Works as expected
> - [@astro/node](https://docs.astro.build/en/guides/integrations-guide/node/): ✅ Works as expected
> - [@astro/netlify](https://docs.astro.build/en/guides/integrations-guide/netlify/): 🟡 Requires Attention
> - [@astro/vercel](https://docs.astro.build/en/guides/integrations-guide/vercel/): 🟡 Requires Attention

Reviewer: ```suggestion
It's recommended to choose one of the [official adapters](https://docs.astro.build/en/guides/server-side-rendering/#official-adapters) Astro offers. While the Node.js and Cloudflare adapters work out-of-the-box with Clerk, the Netlify and Vercel adapters have caveats. See the following sections for more information.
```

### Netlify

> **Development Instances:** Due to Netlify's caching strategies, one of Clerk's core mechanism is unable to work as expected, resulting in infinite redirects that cause an app to fail. This also affects preview environments that use development API keys. We recommend using [production API keys with a new domain.](/docs/deployments/set-up-preview-environment#acquire-an-additional-root-domain)
> 
> Additionally, read [this section](#issues-with-edge-middleware) about **Edge Middleware**

Reviewer: ```suggestion
Due to Netlify's caching strategies, one of Clerk's core mechanisms is unable to work as expected, resulting in infinite redirects that cause an app to fail. This also affects preview environments that use **development** API keys. It's recommended to use [production API keys with a new domain.](/docs/deployments/set-up-preview-environment#acquire-an-additional-root-domain)

Additionally, read [this section](#issues-with-edge-middleware) about **Edge middleware**.
```

### Vercel

> Read [this section](#issues-with-edge-middleware) about **Edge Middleware**

Reviewer: ```suggestion
Read [this section](#issues-with-edge-middleware) about **Edge middleware**.
```

## Issues with Edge middleware

> The netlify and vercel adapters allow developers to convert the astro middleware to an edge middleware.
> Currently, this is expected to not work for the following two reasons.
> - Our Astro SDK populates `Astro.locals.auth` with a function that cannot be serialized.
> - For developers using Astro and React, SSR would not work as expected as the `clerkMiddleware` uses Async Local Storage, and it's stored value will not persist across the runtimes.

Reviewer: ```suggestion
The Netlify and Vercel adapters allow developers to convert the Astro middleware to an Edge middleware.

Currently, this is expected to not work for the following reasons:
- Our Astro SDK populates `Astro.locals.auth` with a function that cannot be serialized.
- For developers using Astro and React, SSR would not work as expected, as `clerkMiddleware` uses Async Local Storage, and it's stored value will not persist across the runtimes.
```
